securedrop/tests/test_source.py

```diff
@@ -83,14 +86,44 @@ def test_generate_already_logged_in(self):
         self.assertEqual(resp.status_code, 200)
         self.assertIn("because you are already logged in.", resp.data)
 
-    def test_create(self):
+    def test_create_new_source(self):
         with self.client as c:
             resp = c.get('/generate')
             resp = c.post('/create', follow_redirects=True)
             self.assertTrue(session['logged_in'])
             # should be redirected to /lookup
             self.assertIn("Submit Materials", resp.data)
 
+    @patch('source.app.logger.warning')
+    @patch('crypto_util.genrandomid',
+           side_effect=[overly_long_codename, 'short codename'])
+    def test_generate_too_long_codename(self, genrandomid, logger):
+        """Generate a codename that exceeds the maximum codename length"""
+
+        with self.client as c:
+            resp = c.post('/generate')
+            self.assertEqual(resp.status_code, 200)
+
+        logger.assert_called_with(
+            "Generated a source codename that was too long, "
+            "skipping it. This should not happen. "
+            "(Codename='{}')".format(overly_long_codename)
+        )
+
+    @patch('source.app.logger.error')
+    def test_create_duplicate_codename(self, logger):
+        with self.client as c:
+            c.get('/generate')
+
+            # Create a source the first time
+            c.post('/create', follow_redirects=True)
+
+            # Attempt to add the same source
+            c.post('/create', follow_redirects=True)
+            logger.assert_called_once()
+            self.assertIn("Attempt to create a source with duplicate codename",
```

Comment:
The logic on the endpoint this tests makes me a bit uneasy. First, this is the schema:

```sql
CREATE TABLE sources (
	id INTEGER NOT NULL, 
	filesystem_id VARCHAR(96), 
	journalist_designation VARCHAR(255) NOT NULL, 
	flagged BOOLEAN, 
	last_updated DATETIME, 
	pending BOOLEAN, 
	interaction_count INTEGER NOT NULL, 
	PRIMARY KEY (id), 
	UNIQUE (filesystem_id), 
	CHECK (flagged IN (0, 1)), 
	CHECK (pending IN (0, 1))
);
```

Note that the `journalist_designation` is not unique. I think we need to open an issue for this.

Second, once we *do* add that fix, the exception there will catch both errors and treat them as the same which is a little problematic. I also am not convinced it should blindly continue with logging the user in there. I actually think it should be a 500 error (maybe a 400). In the UI, that endpoint should never produce a codename that already exists. In the (astronomically tiny) chance it does, we don't want to log them in as another source. I think it should bail and then make the source retry.

Reply:
Good call 😎  - check out issue #2043 and my comments there re: the uniqueness of `journalist_designation`. 

On your second point, duplicate codenames won't be generated due to the unique constraint set on the `filesystem_id` column (recall that `filesystem_id` is [a hash of `codename`](https://github.com/freedomofpress/securedrop/blob/develop/securedrop/source_app/utils.py#L54)).

Reply:
Yes, that is true, but the code here in `source_app/main.py`:

```python
    @view.route('/create', methods=['POST'])
    def create():
        filesystem_id = crypto_util.hash_codename(session['codename'])

        source = Source(filesystem_id, crypto_util.display_id())
        db_session.add(source)
        try:
            db_session.commit()
        except IntegrityError as e:
            db_session.rollback()
            current_app.logger.error(
                "Attempt to create a source with duplicate codename: %s" %
                (e,))
        else:
            os.mkdir(store.path(filesystem_id))

        session['logged_in'] = True
        return redirect(url_for('.lookup'))
```

Will continue on an `IntegrityError` which means in the small chance a source generates the same codename an existing source, we just log them in. That was what I had minor concern about.

But at the end of the day, this PR improves the existing state of things, and perhaps we should revisit this particular problem in another issue.

Reply:
More notes: this endpoint could check if they're logged in or not, and if they are, it could continue, but if they're not logged in *and* they try to make a duplicate user name, they could bail.

Reply:
Hmm yeah, I hear ya. OK - for defense in depth I'm going to make a followup to `abort(500)` after this event is logged and fix the corresponding tests since it's quick and easy